client/rpc-servers/src/middleware.rs

```diff
@@ -18,46 +18,69 @@
 
 //! Middleware for RPC requests.
 
+use std::time::Instant;
+
 use jsonrpc_core::{
 	Middleware as RequestMiddleware, Metadata,
-	Request, Response, FutureResponse, FutureOutput
-};
-use prometheus_endpoint::{
-	Registry, CounterVec, PrometheusError,
-	Opts, register, U64
+	FutureResponse, FutureOutput
 };
+use prometheus_endpoint::{GaugeVec, HistogramOpts, HistogramVec, Opts, PrometheusError, Registry, U64, register};
 
 use futures::{future::Either, Future};
 
 /// Metrics for RPC middleware
 #[derive(Debug, Clone)]
 pub struct RpcMetrics {
-	rpc_calls: Option<CounterVec<U64>>,
+	rpc_calls: HistogramVec,
+	started: GaugeVec<U64>,
+	finished: GaugeVec<U64>
 }
 
 impl RpcMetrics {
 	/// Create an instance of metrics
-	pub fn new(metrics_registry: Option<&Registry>) -> Result<Self, PrometheusError> {
-		Ok(Self {
-			rpc_calls: metrics_registry.map(|r|
-				register(
-					CounterVec::new(
-						Opts::new(
-							"rpc_calls_total",
-							"Number of rpc calls received",
+	pub fn new(metrics_registry: Option<&Registry>) -> Result<Option<Self>, PrometheusError> {
+		if let Some(r) = metrics_registry {
+			Ok(Some(Self {
+				rpc_calls: register(
+					HistogramVec::new(
+						HistogramOpts::new(
+							"rpc_calls_time",
+							"Total time [ms] of processed RPC calls",
 						),
-						&["protocol"]
+						&["protocol", "method"]
 					)?,
 					r,
-				)
-			).transpose()?,
-		})
+				)?,
+				started: register(
+					GaugeVec::new(
+						Opts::new(
+							"rpc_calls_started",
+							"Number of received RPC calls (unique un-batched requests)"
+						),
+						&["protocol", "method"]
+					)?,
+					r
+				)?,
+				finished: register(
+					GaugeVec::new(
+						Opts::new(
+							"rpc_calls_finished",
+							"Number of processed RPC calls (unique un-batched requests)"
```

Comment:
The middleware here uses `on_call` hook (unlike previously `on_request`), so here we don't deal with Batch/Single requests, but rather with individual calls that were part of that request. So if we have a batch of 10 calls it will not be distinguishable from 10 separate requests. I can add separate metric for batch calls, but I'm not sure exactly if it's that useful:
1. correlating calls with batches will be quite hard with current library design (i.e. shared state between `on_request` and `on_call` hooks)
2. I don't think it's used often in polkadot.js/api
3. I feel like the imposed load should be exactly the same - batch is there just to minimize the number of requests from the client perspective, but on the server side it should be roughly the same.